src/notification/NotificationController.ts

```diff
@@ -0,0 +1,266 @@
+import type { Patch } from 'immer';
+import { nanoid } from 'nanoid';
+
+import { hasProperty } from '../util';
+import { BaseController } from '../BaseControllerV2';
+
+import type { RestrictedControllerMessenger } from '../ControllerMessenger';
+
+/**
+ * @type NotificationState
+ * @property notifications - Stores existing notifications to be shown in the UI
+ */
+export type NotificationState = {
+  notifications: Record<string, Notification>;
+};
+
+export type Notification = {
+  id: string;
+  type: NotificationType;
+  origin: string;
+  date: number;
+  read: boolean;
+  message: string;
+};
+
+export enum NotificationType {
+  Native = 'native',
+  InApp = 'in-app',
+}
+
+export interface NotificationArgs {
+  /**
+   * Enum type to determine notification type.
+   */
+  type: NotificationType;
+
+  /**
+   * A message to show on the notification.
+   */
+  message: string;
+}
+
+const name = 'NotificationController';
+
+export type NotificationStateChange = {
+  type: `${typeof name}:stateChange`;
+  payload: [NotificationState, Patch[]];
+};
+
+export type GetNotificationState = {
+  type: `${typeof name}:getState`;
+  handler: () => NotificationState;
+};
+
+export type ShowNotification = {
+  type: `${typeof name}:show`;
+  handler: NotificationController['show'];
+};
+
+export type DismissNotification = {
+  type: `${typeof name}:dismiss`;
+  handler: NotificationController['dismiss'];
+};
+
+export type MarkNotificationRead = {
+  type: `${typeof name}:markRead`;
+  handler: NotificationController['markRead'];
+};
+
+export type GetNotifications = {
+  type: `${typeof name}:getNotifications`;
+  handler: NotificationController['getNotifications'];
+};
+
+export type GetNotificationCount = {
+  type: `${typeof name}:getCount`;
+  handler: NotificationController['getCount'];
+};
+
+export type GetUnreadNotificationCount = {
+  type: `${typeof name}:getUnreadCount`;
+  handler: NotificationController['getUnreadCount'];
+};
+
+export type ControllerActions =
+  | GetNotificationState
+  | ShowNotification
+  | DismissNotification
+  | MarkNotificationRead
+  | GetNotifications
+  | GetNotificationCount
+  | GetUnreadNotificationCount;
+
+export type NotificationMessenger = RestrictedControllerMessenger<
+  typeof name,
+  ControllerActions,
+  NotificationStateChange,
+  never,
+  never
+>;
+
+const metadata = {
+  notifications: { persist: true, anonymous: false },
+};
+
+const defaultState = {
+  notifications: {},
+};
+
+/**
+ * Controller that handles showing notifications to the user and rate limiting origins
```

Comment:
Updated to remove rate limiting and include storage. It doesn't seem like this does rate limiting anymore
```suggestion
 * Controller that handles storing notifications and showing them to the user
```